src/lightning/fabric/plugins/precision/fp8_transformer_engine.py

```diff
@@ -0,0 +1,109 @@
+    ) -> None:
+        if not torch.cuda.get_device_capability() >= (8, 1):
+            raise NotImplementedError("Your CUDA device does not support fp8 mixed precision.")
+        if not _TRANSFORMER_ENGINE_AVAILABLE:
+            raise ModuleNotFoundError(str(_TRANSFORMER_ENGINE_AVAILABLE))
+        from transformer_engine.common.recipe import DelayedScaling
+
+        if recipe is None:
+            recipe = DelayedScaling()
+        elif isinstance(recipe, dict):
+            if "fp8_format" in recipe:
+                from transformer_engine.common.recipe import Format
+
+                recipe["fp8_format"] = getattr(Format, recipe["fp8_format"])
+            recipe = DelayedScaling(**recipe)
+
+        self.recipe = recipe
+        self.replace_layers = replace_layers
+
+    @contextmanager
+    def forward_context(self) -> Generator[None, None, None]:
+        import transformer_engine.pytorch as te
+
+        with te.fp8_autocast(enabled=True, fp8_recipe=self.recipe):
+            yield
+
+    def convert_module(self, module: torch.nn.Module) -> torch.nn.Module:
+        # avoid converting if any is found. assume the user took care of it
+        if self.replace_layers and not any("transformer_engine" in m.__module__ for m in module.modules()):
+            _convert_layers(module)
+        return module
+        # TODO: should we un-convert on teardown?
+
+
+def _convert_layers(module: torch.nn.Module) -> None:
+    import transformer_engine.pytorch as te
+
+    for name, child in module.named_children():
+        if isinstance(child, torch.nn.Linear):
+            if child.in_features % 16 != 0 or child.out_features % 16 != 0:
+                # https://docs.nvidia.com/deeplearning/transformer-engine/user-guide/examples/fp8_primer.html#FP8-autocasting
+                rank_zero_warn(
+                    "Support for FP8 in the linear layers with `precision='8-mixed'` is currently limited to tensors"
+                    " with shapes where both dimensions are divisible by 16. You might want to add padding to your"
+                    " inputs so that you can fit this criteria."
+                )
+                continue
+            has_bias = child.bias is not None
+            replacement = te.Linear(child.in_features, child.out_features, bias=has_bias)
+            replacement.weight.data = child.weight.data.clone()
+            if has_bias:
+                replacement.bias.data = child.bias.data.clone()
+            setattr(module, name, replacement)
+        elif isinstance(child, torch.nn.LayerNorm):
+            replacement = te.LayerNorm(child.normalized_shape[0], eps=child.eps)
+            replacement.weight.data = child.weight.data.clone()
+            replacement.bias.data = child.bias.data.clone()
+            setattr(module, name, replacement)
+        else:
+            # there are other transformer engine layers that we could convert but are more niche. full list at:
```

Comment:
For example, theres a LayerNormMLP class which would replace https://github.com/Lightning-AI/lit-parrot/blob/5594298ce5c1091b5665c4020a8d0bb0adc64949/lit_parrot/model.py#L74-L75 but that would require more complex conversion logic

Reply:
does it make sense to add an additional hook just for this callback, users could call to convert these things manually, so that they don't need to implement/copy all of the above? Especially if we overwrite instance and subclass checks, this is nothing users could rely on anymore...

Reply:
The user controls the model, so they could just copy paste it with the layers replaced.

For single layers conversion, we can easily add them here. But for combinations of layers that's where we cannot do it